pkg/strategy/streaming.go

```diff
@@ -15,84 +17,235 @@ import (
-			c.cronJobs = append(c.cronJobs, *cronjob.CreateEsIndexCleaner(jaeger))
+			manifest.cronJobs = append(manifest.cronJobs, *cronjob.CreateEsIndexCleaner(jaeger))
 		} else {
 			jaeger.Logger().WithField("type", jaeger.Spec.Storage.Type).Warn("Skipping Elasticsearch index cleaner job due to unsupported storage.")
 		}
 	}
 
-	c.dependencies = storage.Dependencies(jaeger)
+	manifest.dependencies = storage.Dependencies(jaeger)
+
+	return manifest
+}
+
+func autoProvisionKafka(jaeger *v1.Jaeger, manifest S) S {
+	if jaeger.Annotations == nil {
+		jaeger.Annotations = map[string]string{}
+	}
+	// mark that we auto provisioned a kafka for this instance
+	jaeger.Annotations[v1.AnnotationProvisionedKafkaKey] = v1.AnnotationProvisionedKafkaValue
+
+	k := kafka.Persistent(jaeger)
+	ku := kafka.User(jaeger)
+	manifest.kafkas = append(manifest.kafkas, k)
+	manifest.kafkaUsers = append(manifest.kafkaUsers, ku)
+
+	// these are the in-container paths, available to the Jaeger containers (ingester/collector)
+	clusterCAPath := fmt.Sprintf("/var/run/secrets/%s-cluster-ca", jaeger.Name)
+	clientCertPath := fmt.Sprintf("/var/run/secrets/%s", ku.Name)
+
+	// this is the volume containing the client TLS details, like the cert and key
+	kuVolume := corev1.Volume{
+		Name: ku.Name,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: ku.Name,
+			},
+		},
+	}
+	// this is the volume containing the CA cluster cert
+	kuCAVolume := corev1.Volume{
+		Name: fmt.Sprintf("%s-cluster-ca", jaeger.Name), // the cluster name is the jaeger name
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: fmt.Sprintf("%s-cluster-ca-cert", jaeger.Name),
+			},
+		},
+	}
+	jaeger.Spec.Volumes = replaceVolume(jaeger.Spec.Volumes, kuVolume, kuCAVolume)
+
+	// and finally, the mount paths to have the secrets in the container file system
+	kuVolumeMount := corev1.VolumeMount{
+		Name:      ku.Name,
+		MountPath: clientCertPath,
+	}
+	kuCAVolumeMount := corev1.VolumeMount{
+		Name:      fmt.Sprintf("%s-cluster-ca", jaeger.Name), // the cluster name is the jaeger name
+		MountPath: clusterCAPath,
+	}
+	jaeger.Spec.VolumeMounts = replaceVolumeMount(jaeger.Spec.VolumeMounts, kuVolumeMount, kuCAVolumeMount)
+
+	brokers := fmt.Sprintf("%s-kafka-bootstrap.kafka.svc.cluster.local:9093", k.Name)
+
+	collectorOpts := asOptionsMap(jaeger.Spec.Collector.Options.Map())
+	ingesterOpts := asOptionsMap(jaeger.Spec.Ingester.Options.Map())
+
+	collectorOpts["kafka.producer.brokers"] = brokers
+	collectorOpts["kafka.producer.authentication"] = "tls"
+	collectorOpts["kafka.producer.tls.ca"] = fmt.Sprintf("%s/ca.crt", clusterCAPath)
+	collectorOpts["kafka.producer.tls.cert"] = fmt.Sprintf("%s/user.crt", clientCertPath)
+	collectorOpts["kafka.producer.tls.key"] = fmt.Sprintf("%s/user.key", clientCertPath)
+	if _, found := collectorOpts["kafka.producer.topic"]; !found {
```

Comment:
Based on the previous issue with kafka deadlock detection interval - shouldn't we leave this default for the backend images? Same applies for consumer.

Reply:
You mean, setting `ingester.deadlockInterval` to `0` when auto-provisioning? I'm not sure we should do it, as the default upstream will be `0` when the auto provisioning becomes available in the Operator 1.15.0. It might make a difference for Jaeger versions pre-1.14, but I think we shouldn't override defaults.

Reply:
Sorry I complicated the issue by talking about deadlockInterval.

However you did make my point for me `... but I think we shouldn't override defaults`. What I was trying to say is the operator shouldn't explicitly be defining what is already a default in the backend already, i.e. `jaeger-spans`.

Reply:
I didn't realize that the default for the topic is `jaeger-spans`. I agree with you then.